source/common/http/filter/squash_filter.cc

```diff
@@ -0,0 +1,319 @@
+  attachment_timeout_timer_->enableTimer(config_->attachmentTimeout());
+  // Check if the timer expired inline.
+  if (state_ == State::NotSquashing) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  return FilterHeadersStatus::StopIteration;
+}
+
+FilterDataStatus SquashFilter::decodeData(Buffer::Instance&, bool) {
+  if (state_ == State::NotSquashing) {
+    return FilterDataStatus::Continue;
+  } else {
+    return FilterDataStatus::StopIterationAndBuffer;
+  }
+}
+
+FilterTrailersStatus SquashFilter::decodeTrailers(HeaderMap&) {
+  if (state_ == State::NotSquashing) {
+    return FilterTrailersStatus::Continue;
+  } else {
+    return FilterTrailersStatus::StopIteration;
+  }
+}
+
+void SquashFilter::setDecoderFilterCallbacks(StreamDecoderFilterCallbacks& callbacks) {
+  decoder_callbacks_ = &callbacks;
+}
+
+void SquashFilter::onCreateAttachmentSuccess(MessagePtr&& m) {
+  in_flight_request_ = nullptr;
+
+  // Get the config object that was created
+  if (Utility::getResponseStatus(m->headers()) != enumToInt(Code::Created)) {
+    ENVOY_LOG(info, "Squash: can't create attachment object. status {} - not squashing",
+              m->headers().Status()->value().c_str());
+    doneSquashing();
+  } else {
+    std::string debugAttachmentId;
+    try {
+      Json::ObjectSharedPtr json_config = getJsonBody(std::move(m));
+      debugAttachmentId = json_config->getObject("metadata", true)->getString("name", EMPTY_STRING);
+    } catch (Json::Exception&) {
+      debugAttachmentId = EMPTY_STRING;
+    }
+
+    if (debugAttachmentId.empty()) {
+      ENVOY_LOG(info, "Squash: failed to parse debug attachment object - check server settings.");
+      doneSquashing();
+    } else {
+      debugAttachmentPath_ = POST_ATTACHMENT_PATH + debugAttachmentId;
+      pollForAttachment();
+    }
+  }
+}
+
+void SquashFilter::onCreateAttachmentFailure(AsyncClient::FailureReason) {
+  // in_flight_request_ will be null if we are called inline of async client send()
+  bool request_created = in_flight_request_ != nullptr;
+  in_flight_request_ = nullptr;
+
+  // No retries here, as we couldnt create the attachment object.
+  if (request_created) {
+    // Cleanup not needed if onFailure called inline in async client send, as this means that
+    // decodeHeaders is down the stack and will return Continue.
+    doneSquashing();
+  }
+}
+
+void SquashFilter::onGetAttachmentSuccess(MessagePtr&& m) {
+  in_flight_request_ = nullptr;
+
+  std::string attachmentstate;
+  try {
+    Json::ObjectSharedPtr json_config = getJsonBody(std::move(m));
+    attachmentstate = json_config->getObject("status", true)->getString("state", EMPTY_STRING);
+  } catch (Json::Exception&) {
+    // No state yet.. leave it empty for the retry logic.
+  }
+
+  bool attached = attachmentstate == ATTACHED_STATE;
+  bool error = attachmentstate == ERROR_STATE;
+  bool finalstate = attached || error;
+
+  if (finalstate) {
+    doneSquashing();
+  } else {
+    scheduleRetry();
```

Comment:
Mind adding a comment that the filter will continue trying until attachment_timeout_timer_ expires?
I forgot there were two timers for a moment and thought this might infinite loop :-P